kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -804,13 +842,186 @@ impl<'tcx> GotocCtx<'tcx> {
+                fields.push(DatatypeComponent::Field {
+                    name: ctx.generator_variant_name(var_idx),
+                    typ: ctx.codegen_generator_variant_struct(
+                        generator_name,
+                        pretty_name,
+                        type_and_layout.for_variant(ctx, var_idx),
+                        variant_name,
+                        None,
+                    ),
+                });
+            }
+            fields
+        })
+    }
+
+    /// Generates a struct for a variant of the generator.
+    ///
+    /// The parameter `discriminant_field` may contain the index of a field that is the discriminant.
+    fn codegen_generator_variant_struct(
+        &mut self,
+        generator_name: InternedString,
+        pretty_generator_name: InternedString,
+        type_and_layout: TyAndLayout<'tcx, Ty<'tcx>>,
+        variant_name: InternedString,
+        discriminant_field: Option<usize>,
+    ) -> Type {
+        let struct_name = format!("{generator_name}::{variant_name}");
+        let pretty_struct_name = format!("{pretty_generator_name}::{variant_name}");
+        debug!(?pretty_struct_name, "codeged_generator_variant_struct");
+        self.ensure_struct(struct_name, pretty_struct_name, |ctx, _| {
+            let mut offset = Size::ZERO;
+            let mut fields = vec![];
+            for idx in type_and_layout.fields.index_by_increasing_offset() {
+                // The discriminant field needs to have a special name to work with the rest of the code:
+                let field_name = if Some(idx) == discriminant_field {
+                    "case".into()
+                } else {
+                    ctx.generator_field_name(idx).into()
+                };
+                let field_ty = type_and_layout.field(ctx, idx).ty;
+                let field_offset = type_and_layout.fields.offset(idx);
+                let field_size = type_and_layout.field(ctx, idx).size;
+                if let Some(padding) =
+                    ctx.codegen_struct_padding(offset.bits(), field_offset.bits(), idx)
+                {
+                    fields.push(padding);
+                }
+                fields.push(DatatypeComponent::Field {
+                    name: field_name,
+                    typ: ctx.codegen_ty(field_ty),
+                });
+                offset += field_size;
+            }
+            // Add padding to ensure that the size of the struct is a multiple of the alignment
+            let align = type_and_layout.layout.align().abi.bits();
+            let overhang = offset.bits() % align;
+            if overhang != 0 {
+                fields.push(
+                    ctx.codegen_struct_padding(
+                        offset.bits(),
+                        offset.bits() + align - overhang,
+                        type_and_layout.fields.count(),
+                    )
+                    .unwrap(),
+                )
+            }
```

Comment:
Merge this with the enum code as a utility fn

Reply:
Done